lib/propolis/src/cpuid.rs

```diff
@@ -0,0 +1,413 @@
+    }
+
+    /// Specify CPU vendor
+    pub fn with_vendor(self, vendor: VendorKind) -> Self {
+        Self { vendor_kind: Some(vendor), ..self }
+    }
+
+    /// Specify CPU topology types to render into the specialized [Set]
+    pub fn with_cpu_topo(self, topos: impl Iterator<Item = TopoKind>) -> Self {
+        let mut cpu_topo = BTreeSet::new();
+        for t in topos {
+            cpu_topo.insert(t);
+        }
+
+        Self { cpu_topo, ..self }
+    }
+
+    /// Update cache topology information for specified vCPU count and SMT
+    /// capabilities
+    pub fn with_cache_topo(self) -> Self {
+        Self { do_cache_topo: true, ..self }
+    }
+
+    /// Given the attributes and modifiers specified in this [Specializer],
+    /// render an updated [Set] reflecting those data.
+    pub fn execute(self, mut set: Set) -> Set {
+        // Use vendor override if provided, or else the existing one
+        if let Some(vendor) = self.vendor_kind {
+            set.vendor = vendor;
+        }
+        match set.vendor {
+            VendorKind::Amd => {
+                if self.do_cache_topo && self.num_vcpu.is_some() {
+                    self.fix_amd_cache_topo(&mut set);
+                }
+            }
+            _ => {}
+        }
+
+        // apply any requested topo info fixups
+        self.fix_cpu_topo(&mut set);
+
+        // APIC ID based on vcpuid
+        if let Some(vcpuid) = self.vcpuid.as_ref() {
+            if let Some(ent) = set.get_mut(Ident(0x1, None)) {
+                // bits 31:24 contain initial APIC ID
+                ent.ebx &= !0xff000000;
+                ent.ebx |= ((*vcpuid as u32) & 0xff) << 24;
+            }
+        }
+
+        // logical CPU count (if SMT is enabled)
+        if let Some(num_vcpu) = self.num_vcpu.as_ref() {
+            if self.has_smt {
+                if let Some(ent) = set.get_mut(Ident(0x1, None)) {
+                    ent.edx |= (0x1 << 28);
+                    // bits 23:16 contain max IDs for logical CPUs in package
+                    ent.ebx &= !0xff0000;
+                    ent.ebx |= (num_vcpu.get() as u32) << 16;
+                }
+            }
+        }
+
+        set
+    }
+
+    fn fix_amd_cache_topo(&self, set: &mut Set) {
+        assert!(self.do_cache_topo);
+        let num = self.num_vcpu.unwrap().get();
+        for ecx in 0..u32::MAX {
+            match set.get_mut(Ident(0x8000001d, Some(ecx))) {
+                None => break,
+                Some(vals) => {
+                    // bits 7:5 hold the cache level
+                    let visible_count = match (vals.eax & 0b11100000 >> 5) {
+                        0b001 | 0b010 => {
+                            // L1/L2 shared by SMT siblings
+                            if self.has_smt {
+                                2
+                            } else {
+                                1
+                            }
+                        }
+                        0b011 => {
```

Comment:
This is fine for now but is a little different from the RFD 314 proposal here (I've leaned toward exposing the underlying CCX membership to guests so that guest schedulers can cooperate with whatever CCX-aware scheduling we happen to implement). We should file an issue to close on this.

As with the B/1F/8..1E settings, this is a place where I think it might make sense to honor explicit settings from the caller if they happen to exist (so that the control plane can define distinct virtual platforms that expose different cache topologies and apply them at different times to different instances without having to upgrade/downgrade any Propolis binaries).

Reply:
Until we have better core/thread-aware scheduling, I think going as far as CCX topology seems a bit... much?  I thought I had this close to matching what 314 spelled out.

Reply:
Section 6.4.13 of the RFD gets into this a bit, though embarrassingly for me it has some unresolved comments that I really need to get back to... Basically my view is that we can either

1. reflect the CCX topology we'd like to have now, so that if/when we have scheduling enlightenments that actually schedule vCPUs according to their vCCX topology, the enlightenments "light up" underneath running VMs without requiring them to restart so we can change guest CPUID
2. put everyone in the same NUMA domain into a single virtual unified cache, in the hopes that this will save guest schedulers from doing useless work to account for L3 locality (but at the cost of requiring a VM to stop and start to pick up new CPUID settings later)

In the RFD I favored option 1, but I don't feel very strongly about it at the moment, since we have relatively few production VMs right now and aren't currently updating them without stopping them first. And we still have significant runway to make policy changes here. So I don't think we have to change anything in this PR; mostly I think I need to revisit 314 and make sure we get to ground on whether we're eventually going to do this.

(With all that said, I don't think we have to do a full-blown core scheduler to have some level of L3 locality--if we end up pinning vCPUs to host LPs then it would be at least somewhat straightforward to try to be CCX-aware when we assign those mappings.)

Reply:
> I don't think we have to do a full-blown core scheduler to have some level of L3 locality--if we end up pinning vCPUs to host LPs then it would be at least somewhat straightforward to try to be CCX-aware when we assign those mappings.

But which further complicates the bin-packing problem.  Since we're likely to revise the CPUID data when we add other HVM enlightenments, I'm inclined to go with a flat NUMA domain for now, knowing we'll exposing a more sophisticated model later.